[SwiftUI] Create cross-import SwiftUI overlay

rr-codes opened the pull request:
#### 4885f986934d30f1d606cc43739c48676f680442
<pre>
[SwiftUI] Create cross-import SwiftUI overlay
<a href="https://bugs.webkit.org/show_bug.cgi?id=283455">https://bugs.webkit.org/show_bug.cgi?id=283455</a>
<a href="https://rdar.apple.com/140050834">rdar://140050834</a>

Reviewed by Elliott Williams.

Add a new `_WebKit_SwiftUI` target to support clients not always needing to import SwiftUI when importing WebKit.
This also serves as a fix to being able to build WebKit in certain configurations.

* Source/WebKit/Configurations/_WebKit_SwiftUI.xcconfig: Added.
* Source/WebKit/Modules/SwiftUI.swiftoverlay: Added.
* Source/WebKit/UIProcess/API/Swift/WebPage+Navigation.swift:
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(WebPage_v0.url):
(WebPage_v0.title):
(WebPage_v0.estimatedProgress):
(WebPage_v0.isLoading):
(WebPage_v0.serverTrust):
(WebPage_v0.hasOnlySecureContent):
(WebPage_v0.backingWebView):
(KeyValueObservations.contents):
(WebPage_v0.themeColor): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/_WebKit_SwiftUI/CrossImportOverlay.swift: Added.
* Source/WebKit/_WebKit_SwiftUI/Info.plist: Added.
* Source/WebKit/_WebKit_SwiftUI/WebPage+SwiftUI.swift: Added.
(WebPage_v0.themeColor):
* Source/WebKit/_WebKit_SwiftUI/WebView.swift: Renamed from Source/WebKit/UIProcess/API/Swift/WebView.swift.
(WebView_v0.body):
(WebViewWrapper.frame):
(WebViewWrapper.bounds):
(WebViewWrapper.webView):
(WebViewWrapper.updateWebViewFrame):
(WebViewRepresentable.makePlatformView(_:)):
(WebViewRepresentable.updatePlatformView(_:context:)):
(WebViewRepresentable.makeUIView(_:)):
(WebViewRepresentable.updateUIView(_:context:)):
(WebViewRepresentable.makeNSView(_:)):
(WebViewRepresentable.updateNSView(_:context:)):
* Tools/MiniBrowser/SwiftUI/ContentView.swift:
* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPage.swift:

Canonical link: <a href="https://commits.webkit.org/289906@main">https://commits.webkit.org/289906@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ad9a850c58cf3ba261600c00d673b11ce6d09788

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88215 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7732 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42656 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93168 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38965 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90266 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8116 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15913 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68056 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25780 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91217 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6199 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79805 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48424 "Found 1 new API test failure: /TestJSC:/jsc/options (failure)") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5971 "Found 9 new test failures: imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-006.html imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml.html imported/w3c/web-platform-tests/html/semantics/disabled-elements/event-propagate-disabled-keyboard.tentative.html imported/w3c/web-platform-tests/webaudio/the-audio-api/the-audiocontext-interface/audiocontext-getoutputtimestamp-cross-realm.html imported/w3c/web-platform-tests/webaudio/the-audio-api/the-audioworklet-interface/audioworkletprocessor-process-zero-outputs.https.html imported/w3c/web-platform-tests/webaudio/the-audio-api/the-audioworklet-interface/audioworkletprocessor-promises.https.html imported/w3c/web-platform-tests/webaudio/the-audio-api/the-channelmergernode-interface/active-processing.https.html imported/w3c/web-platform-tests/webaudio/the-audio-api/the-convolvernode-interface/active-processing.https.html imported/w3c/web-platform-tests/webaudio/the-audio-api/the-convolvernode-interface/realtime-conv.html (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/34215 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38073 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76357 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35099 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95014 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15385 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11282 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76916 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15641 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75662 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76160 "Found 1 new API test failure: /TestJSC:/jsc/options (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20552 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18983 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8412 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13809 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15403 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20702 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15145 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18591 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16927 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

davidquesada commented on Source/WebKit/Configurations/_WebKit_SwiftUI.xcconfig:
```diff
@@ -0,0 +1,72 @@
+PRODUCT_BUNDLE_IDENTIFIER = com.apple.WebKit.$(PRODUCT_NAME:rfc1034identifier);
+INFOPLIST_FILE = Info.plist;
+
+INSTALL_PATH = $(WEBKIT_FRAMEWORKS_DIR);
```
If this is going to be installed next to WebKit.framework in WEBKIT_FRAMEWORKS_DIR, you'll need to add a "Create Symlink to Alt Root Path" script phase like WebKit.framework does. That adds the symlink in /System/Library/Frameworks to the real path in the OS cryptex, which allows clients to load the framework from the SDK.

(As an aside, I've been thinking about going in and renaming all the "alternate root" stuff to more explicitly refer to the OS cryptex by name, since we don't need the extra obfuscation for secrecy anymore, and that just is asking to be confused with things like WK_ALTERNATE_FRAMEWORKS_DIR, which refers to Mac Catalyst. Maybe one of these days.)

davidquesada commented on Source/WebKit/Configurations/_WebKit_SwiftUI.xcconfig:
```diff
@@ -0,0 +1,72 @@
+ENABLE_SWIFTUI[sdk=iphone*17*] = NO;
+ENABLE_SWIFTUI[sdk=watch*] = NO;
+ENABLE_SWIFTUI[sdk=appletv*] = NO;
+ENABLE_SWIFTUI[sdk=xr*1*] = NO;
```
I wonder if it would be better to explicitly add rules for `[sdk=xros1.*]` and `[sdk=xrsimulator1.*]` here, since `xr*1*` can match things like `xros2.1`. See rdar://131234802

axiixc commented on Source/WebKit/UIProcess/API/Swift/View+WebViewModifiers.swift:
```diff
@@ -27,15 +27,25 @@ import Foundation
 extension EnvironmentValues {
-    @Entry var webViewAllowsBackForwardNavigationGestures = false
+    @Entry
+    @_spi(CrossImportOverlay)
```
What's this doing? I would expect us to move any files with imports of `SwiftUI` to the new target wholesale, or will that just come later?

axiixc reviewed (commented):
In broad strokes this seems good, although it's unclear if this is intended as our final state or just an initial part with followup? I'd also defer to someone else on some of the project and build system changes.

emw-apple commented on Source/WebKit/Configurations/_WebKit_SwiftUI.xcconfig:
```diff
@@ -0,0 +1,80 @@
+ENABLE_SWIFTUI[sdk=watch*] = NO;
+ENABLE_SWIFTUI[sdk=appletv*] = NO;
+ENABLE_SWIFTUI[sdk=xros1*] = NO;
+ENABLE_SWIFTUI[sdk=xrsimulator1*] = NO;
```
What is this setting used-for, is it built in? If it's a WebKit thing it should probably have a `WK_` prefix. I also wonder if you can de-dupe this with your definition of SWIFT_ACTIVE_COMPILATION_CONDITIONS above.

rr-codes replied:
it's a WK thing, so that the files are basically ignored via SKIP_INSTALL and EXCLUDED_SOURCE_FILE_NAMES

emw-apple commented on Source/WebKit/Configurations/_WebKit_SwiftUI.xcconfig:
```diff
@@ -0,0 +1,80 @@
+EXCLUDED_SOURCE_FILE_NAMES = $(EXCLUDED_SOURCE_FILE_NAMES_$(ENABLE_SWIFTUI));
+EXCLUDED_SOURCE_FILE_NAMES_NO = *;
+
+INCLUDED_SOURCE_FILE_NAMES = _WebKit_SwiftUI/Empty.swift;
```
I assume this is needed to support building without ENABLE_SWIFTUI. Can you leave a comment that the file can be removed when all configurations build WebKit's SwiftUI support?

rr-codes replied:
yup, will do

emw-apple commented on Source/WebKit/Configurations/_WebKit_SwiftUI.xcconfig:
```diff
@@ -0,0 +1,80 @@
+
+INCLUDED_SOURCE_FILE_NAMES = _WebKit_SwiftUI/Empty.swift;
+
+USE_HEADERMAP = NO;
```
Why? We don't do this in our other projects, and I would strongly prefer we avoid it here. Selectively disabling headermaps breaks some expectations made by Xcode and the modules system about how compilation works.

emw-apple commented on Source/WebKit/UIProcess/API/Cocoa/WKPreferencesPrivate.h:
```diff
@@ -236,6 +236,10 @@ typedef NS_ENUM(NSInteger, _WKPitchCorrectionAlgorithm) {
 #endif
 
+#if TARGET_OS_IPHONE
+@property (nonatomic) BOOL tabFocusesLinks WK_API_AVAILABLE(ios(WK_IOS_TBA), visionos(WK_XROS_TBA));
```
Targets that build from the same Xcode project, should be able to share implementation details without exposing them as SPI. Were you able to look into whether the cross-import target can `import WebKit_Internal`? I still totally believe that it can, so if I'm wrong I'd like to understand why.

rr-codes replied:
I forgot to; I will now

rr-codes replied:
I confirmed this is not possible; I get an error "No such module 'WebKit_Internal'"

emw-apple replied:
Sorry to be picky about this, but this isn’t a good enough reason. I believe it’s possible—modules are just files that the compiler has to discover on a search path. And avoiding needless SPI exports is a good goal.

Did you try setting SWIFT_INCLUDE_PATHS to include the path to WebKit/Modules/Internal? WebKit.xcconfig does this. You could probably just move this setting to BaseTarget.xcconfig since we want it to work from other targets:

```
SWIFT_INCLUDE_PATHS = $(SRCROOT)/Modules/Internal;
```

emw-apple commented on Source/WebKit/UIProcess/API/Swift/WebPage+NavigationDeciding.swift:
```diff
@@ -51,7 +51,8 @@ extension WebPage_v0 {
 #endif
 
-        var wrapped: WKNavigationAction
+        @_spi(CrossImportOverlay)
```
Ditto above, though I think this one would be much harder to move into `WebKit_Internal`, because you'd have to get the latter module to have an associated swiftmodule. I think that would mean a new target.

IMO it's fine to keep this as SPI, though annoying that Swift doesn't provide the same flexibility that clang-based languages do.

rr-codes replied:
once WebKit adopts Swift package visibility, this can be switched to package instead of SPI public. I intend to do this later on

emw-apple reviewed (changes requested):
Looks great, though I have a few questions that I'd like to work through:

emw-apple commented on Source/WebKit/Configurations/_WebKit_SwiftUI.xcconfig:
```diff
@@ -0,0 +1,84 @@
+SWIFT_INSTALL_OBJC_HEADER = NO;
+DEFINES_MODULE = NO;
+
+OTHER_SWIFT_FLAGS = -Xfrontend -experimental-spi-only-imports $(OTHER_SWIFT_FLAGS_AVAILABILITY_$(USE_INTERNAL_SDK));
```
We set this in Base.xcconfig, so you shouldn't need it here.

rr-codes replied:
I initially had that, but realized I don't want the `-no-warnings-as-errors` setting since I want all warnings to be errors in this project

emw-apple replied:
Ah, thanks for explaining your intention here.

I don't think that's feasible, because Swift doesn't provide a way to temporarily silence a deprecation or syntax warning (rdar://26611918). That's essential for our project, because we build for multiple OS releases and sometimes have to retain old syntax that has become a warning in recent compilers (e.g. `@_implementationOnly`), and because we need to rapidly respond to deprecations to avoid causing internal build breakage.

Maybe you can get away with warnings as errors today, since this is a new feature, but you won't be able to long term. And regardless, I think that these policies should be set project-wide whenever possible.

emw-apple commented on Source/WebKit/Configurations/_WebKit_SwiftUI.xcconfig:
```diff
@@ -0,0 +1,84 @@
+OTHER_SWIFT_FLAGS = -Xfrontend -experimental-spi-only-imports $(OTHER_SWIFT_FLAGS_AVAILABILITY_$(USE_INTERNAL_SDK));
+
+// Disable optimization for debug builds.
+SWIFT_OPTIMIZATION_LEVEL = -O;
+SWIFT_OPTIMIZATION_LEVEL[config=Debug] = -Onone;
```
I know this is copied from WebKitSwift, but do you know if it's actually needed? It seems weird, like, of course debug builds should be `-Onone` and I would expect the build system to do that by default.

rr-codes replied:
it is needed afaik; Swift Previews complain with an error when this is omitted

axiixc replied:
What we should really do is sink this into the proper `Debug.xcconfig` or whatever files. When I did the overlay work I wanted to keep anything even tangentially Swift-related out of the standard configs (because we weren’t sure if it’d have unwanted effects in B&I), but these days we should just do this normally.

>  I would expect the build system to do that by default.

It’s a reasonable thing to want, but IIRC all options have one singular default (even in C). Xcode doesn’t actually understand “configs”, they’re just named collections of overrides without semantics, so it has no real understand of what “Debug” is. We have to bring that all ourselves.

emw-apple replied:
I checked XCBuild and the default is indeed `-O`. Weird.

emw-apple commented on Source/WebKit/Configurations/_WebKit_SwiftUI.xcconfig:
```diff
@@ -0,0 +1,84 @@
+
+CLANG_INSTRUMENT_FOR_OPTIMIZATION_PROFILING = NO; // Disable PGO profile generation.
+
+OTHER_LDFLAGS = $(inherited) $(SOURCE_VERSION_LDFLAGS) $(FRAMEWORK_LDFLAGS_$(WK_PLATFORM_NAME));
```
FRAMEWORK_LDFLAGS_* looks like a holdover from WebKitSwift, please remove.
```suggestion
OTHER_LDFLAGS = $(inherited) $(SOURCE_VERSION_LDFLAGS);
```

emw-apple commented on Tools/SwiftBrowser/Source/ContentView.swift:
```diff
@@ -24,6 +24,7 @@
 import SwiftUI
 import UniformTypeIdentifiers
 @_spi(Private) import WebKit
+@_spi(Private) import _WebKit_SwiftUI
```
What does this do? Seems like we should already get the cross-import in this file, because of the SwiftUI and WebKit imports. And the cross-import module does not itself have any `@_spi(Private)` declared.

rr-codes replied:
when using SPI, it's necessary to use the cross-import name explicitly. And _WebKit_SwiftUI does have several `@_spi(Private)` symbols currently (such as webViewAllowsBackForwardNavigationGestures)

emw-apple replied:
I see, it was hiding outside the diff. Thanks!

emw-apple commented on Tools/SwiftBrowser/SwiftBrowser.xcodeproj/project.pbxproj:
```diff
@@ -18,8 +20,23 @@
+			runOnlyForDeploymentPostprocessing = 0;
+		};
+/* End PBXCopyFilesBuildPhase section */
+
```
I don't think this is needed—did you intentionally add it, or is it just the default "Embed" phase that Xcode created? We don't intend to ship SwiftBrowser.app as a standalone application, so it does not need to bundle the frameworks it's using (and if it did, we'd likely want to bundle the entire WebKit stack).

rr-codes replied:
I don't recall intentionally adding this. Where in Xcode UI would I remove this?

emw-apple replied:
SwiftBrowser.xcodeproj > SwiftBrowser > Build Phases. It probably got added when you told it to link against _WebKit_SwiftUI.framework (and presumably _that_ was intentional, for build ordering?)

emw-apple commented on Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageTests.swift:
```diff
@@ -23,9 +23,11 @@
 import Observation
 import Testing
 @_spi(Private) @_spi(Testing) import WebKit
+@_spi(Private) import _WebKit_SwiftUI
```
Ditto, what does this get us?

rr-codes replied:
the same thing as https://github.com/WebKit/WebKit/pull/37452#discussion_r1890967519

emw-apple commented on Tools/SwiftBrowser/SwiftBrowser.xcodeproj/project.pbxproj:
```diff
@@ -18,8 +19,22 @@
 		07A58C4A2D0567A100CAB4ED /* SettingsView.swift in Sources */ = {isa = PBXBuildFile; fileRef = 07A58C492D0567A100CAB4ED /* SettingsView.swift */; };
 /* End PBXBuildFile section */
 
+/* Begin PBXCopyFilesBuildPhase section */
+		07A4CEC52D092C0800764F5E /* Embed Frameworks */ = {
+			isa = PBXCopyFilesBuildPhase;
+			buildActionMask = 2147483647;
+			dstPath = "";
+			dstSubfolderSpec = 10;
+			files = (
+			);
+			name = "Embed Frameworks";
+			runOnlyForDeploymentPostprocessing = 0;
+		};
+/* End PBXCopyFilesBuildPhase section */
+
```
Please remove the whole Embed Frameworks build phase, since it isn't doing anything.

emw-apple reviewed (approved):
r+ with one minor change. And please be sure to have tested this on internal builds of all platforms before you merge, since it's a major build system change.

emw-apple commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -19611,6 +19727,24 @@
+			isa = PBXSourcesBuildPhase;
+			buildActionMask = 2147483647;
+			files = (
+				07275C692D00E7DF002315A5 /* (null) in Sources */,
```
Looks like cruft for a file that got deleted, please remove.

emw-apple commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -28,6 +28,7 @@
 				2DA8152C28A4204400CF811C /* PBXTargetDependency */,
 				BC8283D516B4C01F00A278FE /* PBXTargetDependency */,
 				2D9FB22C2375245C0049F936 /* PBXTargetDependency */,
+				07275C6B2D00ED43002315A5 /* PBXTargetDependency */,
```
This aggregate target is what gets built by Apple's internal build system. You can't list your target here without causing overlapping content. Instead, put it in the "All" aggregate target like we do for WebKitSwift.

rr-codes committed:
[SwiftUI] Create cross-import SwiftUI overlay
https://bugs.webkit.org/show_bug.cgi?id=283455
rdar://140050834

Reviewed by Elliott Williams.

Add a new `_WebKit_SwiftUI` target to support clients not always needing to import SwiftUI when importing WebKit.
This also serves as a fix to being able to build WebKit in certain configurations.

* Source/WebKit/Configurations/_WebKit_SwiftUI.xcconfig: Added.
* Source/WebKit/Modules/SwiftUI.swiftoverlay: Added.
* Source/WebKit/UIProcess/API/Swift/WebPage+Navigation.swift:
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(WebPage_v0.url):
(WebPage_v0.title):
(WebPage_v0.estimatedProgress):
(WebPage_v0.isLoading):
(WebPage_v0.serverTrust):
(WebPage_v0.hasOnlySecureContent):
(WebPage_v0.backingWebView):
(KeyValueObservations.contents):
(WebPage_v0.themeColor): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/_WebKit_SwiftUI/CrossImportOverlay.swift: Added.
* Source/WebKit/_WebKit_SwiftUI/Info.plist: Added.
* Source/WebKit/_WebKit_SwiftUI/WebPage+SwiftUI.swift: Added.
(WebPage_v0.themeColor):
* Source/WebKit/_WebKit_SwiftUI/WebView.swift: Renamed from Source/WebKit/UIProcess/API/Swift/WebView.swift.
(WebView_v0.body):
(WebViewWrapper.frame):
(WebViewWrapper.bounds):
(WebViewWrapper.webView):
(WebViewWrapper.updateWebViewFrame):
(WebViewRepresentable.makePlatformView(_:)):
(WebViewRepresentable.updatePlatformView(_:context:)):
(WebViewRepresentable.makeUIView(_:)):
(WebViewRepresentable.updateUIView(_:context:)):
(WebViewRepresentable.makeNSView(_:)):
(WebViewRepresentable.updateNSView(_:context:)):
* Tools/MiniBrowser/SwiftUI/ContentView.swift:
* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPage.swift:

Canonical link: https://commits.webkit.org/289906@main
